### PR TITLE
Warn that a user-defined API on an inline array type won’t be used by the language for an element access expression

### LINKED ISSUE #68868: Surface a warning at the declaration site for members that will be hidden by the language `[InlineArray]` support
### Summary

The new `[InlineArray]` support adds some implicit functionality for the attributed types. This implicit functionality is not defined as an actual API and can mask any explicit API surface the user has defined themselves.

For example, the following program:
```csharp
using System.Diagnostics.CodeAnalysis;
using System.Runtime.CompilerServices;

S1 s1 = default;
_ = s1[0];
_ = s1[(nuint)0];

[InlineArray(10)]
public struct S1
{
    public int _x;

    [UnscopedRef]
    public ref int this[int index]
    {
        get
        {
            Console.WriteLine("int indexer");
            return ref Unsafe.Add(ref Unsafe.As<S1, int>(ref this), index);
        }
    }

    [UnscopedRef]
    public ref int this[nuint index]
    {
        get
        {
            Console.WriteLine("nuint indexer");
            return ref Unsafe.Add(ref Unsafe.As<S1, int>(ref this), index);
        }
    }
}
```

Outputs:
```
nuint indexer
```

The expected output from the user might have been:
```
int indexer
nuint indexer
```

### Expected Behavior

The compiler should surface a warning indicating that the member is now hidden and inaccessible:
```
// ...
    [UnscopedRef]
    public ref int this[int index] // CS----: Inline array member is inaccessible in C# 12 or later
    {
// ...
```

A warning, rather than an error, is emitted to ensure that users are not forced to break binary compatibility if they are migrating an existing type to use `[InlineArray]`. 

### Alternative Designs

The language could also actually define a public indexer for the various types it supports and then surface `CS0111: Type 'S1' already defines a member called 'this' with the same parameter types`. However, this would increase the size of the type and could lead to "less ideal" IL emitted since it would be one more thing the JIT has to process and potentially inline.

Relates to test plan https://github.com/dotnet/roslyn/issues/67826

### PR DESCRIPTION
**Pull request by AlekseyTs**
Closes #68868.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
CC @tannergooding

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs [hunk 1]
```diff
@@ -1758,6 +1758,50 @@ protected override void AfterMembersCompletedChecks(BindingDiagnosticBag diagnos
...
+                        foreach (var conversion in GetMembers().OfType<SourceUserDefinedConversionSymbol>())
+                        {
+                            TypeSymbol returnType = conversion.ReturnType;
+                            TypeSymbol returnTypeOriginalDefinition = returnType.OriginalDefinition;
+
+                            if (conversion.ParameterCount == 1 &&
+                                conversion.Parameters[0].Type.Equals(this, TypeCompareKind.AllIgnoreOptions) &&
+                                (returnTypeOriginalDefinition.Equals(span ??= DeclaringCompilation.GetWellKnownType(WellKnownType.System_Span_T), TypeCompareKind.AllIgnoreOptions) ||
+                                 returnTypeOriginalDefinition.Equals(readOnlySpan ??= DeclaringCompilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T), TypeCompareKind.AllIgnoreOptions)) &&
+                                Conversions.HasIdentityConversion(((NamedTypeSymbol)returnTypeOriginalDefinition).Construct(ImmutableArray.Create(elementType)), returnType))
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs:1799]: > What's this check for?

This checks that the return type of the conversion is indeed something that conflicts with built-in conversions.
For example, in ```UserDefinedConversion_Warning_03``` it is not.

File: src/Compilers/CSharp/Test/Emit2/Semantics/InlineArrayTests.cs [hunk 2]
```diff
@@ -20696,5 +20722,407 @@ .locals init (int V_0)
...
+        public void UserDefinedSlice_Warning_01()
+        {
+            var src = @"
+[System.Runtime.CompilerServices.InlineArray(4)]
+struct Buffer4
+{
+    private int _element0;
+
+    int Length => 4;
+    string Slice(int i, int j) => ""int"";
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit2/Semantics/InlineArrayTests.cs:20888]: > Consider testing explicit interface implementation of Slice

Added UserDefinedSlice_Warning_03

**Review by jjonescz — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler For the second review

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler For the second review

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler For the second review

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler For the second review

**Review by cston — APPROVED** [review]

**Pull request merged**